src/lsm/scan_fuzz.zig

```diff
@@ -0,0 +1,991 @@
+            if (value == 0) {
+                // Fill the zeroed fields with random values out of the matching prefix.
+                other.set_index(
+                    index,
+                    prefix_combine(std.math.maxInt(u32), random.int(u32)),
+                );
+            }
+        }
+
+        return other;
+    }
+};
+
+const ThingsGroove = GrooveType(
+    Storage,
+    Thing,
+    .{
+        .ids = .{
+            .id = 1,
+            .index_1 = 2,
+            .index_2 = 3,
+            .index_3 = 4,
+            .index_4 = 5,
+            .index_5 = 6,
+            .index_6 = 7,
+            .index_7 = 8,
+            .index_8 = 9,
+            .index_9 = 10,
+            .index_10 = 11,
+            .index_11 = 12,
+            .index_12 = 13,
+            .timestamp = 14,
+        },
+        .batch_value_count_max = .{
+            .id = batch_max,
+            .index_1 = batch_max,
+            .index_2 = batch_max,
+            .index_3 = batch_max,
+            .index_4 = batch_max,
+            .index_5 = batch_max,
+            .index_6 = batch_max,
+            .index_7 = batch_max,
+            .index_8 = batch_max,
+            .index_9 = batch_max,
+            .index_10 = batch_max,
+            .index_11 = batch_max,
+            .index_12 = batch_max,
+            .timestamp = batch_max,
+        },
+        .ignored = &[_][]const u8{"reserved"},
+        .derived = .{},
+    },
+);
+
+const Forest = ForestType(Storage, .{
+    .things = ThingsGroove,
+});
+
+const Index = std.meta.FieldEnum(ThingsGroove.IndexTrees);
+
+const ScanLookup = ScanLookupType(
+    ThingsGroove,
+    ThingsGroove.ScanBuilder.Scan,
+    Storage,
+);
+
+const QueryCondition = union(enum) {
+    field_condition: FieldCondition,
+    parenthesis_condition: ParenthesisCondition,
+
+    pub fn format(
+        self: QueryCondition,
+        comptime _: []const u8,
+        _: std.fmt.FormatOptions,
+        writer: anytype,
+    ) !void {
+        switch (self) {
+            .field_condition => |field_condition| try writer.print(
+                "{s}={}",
+                .{
+                    std.enums.tagName(Index, field_condition.index).?,
+                    field_condition.value,
+                },
+            ),
+            .parenthesis_condition => |parenthesis_condition| {
+                try writer.print("(", .{});
+                for (parenthesis_condition.operands, 0..) |operand, i| {
+                    try writer.print("{}", .{operand});
+                    if (i < parenthesis_condition.operands.len - 1) {
+                        switch (parenthesis_condition.operator) {
+                            .union_set => try writer.print(" OR ", .{}),
+                            .intersection_set => try writer.print(" AND ", .{}),
+                        }
+                    }
+                }
+                try writer.print(")", .{});
+            },
+        }
+    }
+};
```

Comment:
So this gives me a pause. My nit comment is that, with many types flowing around, it's a bit hard to see the overall recursive structure of of types here. So perhaps some judicial usage of nesting would help to illumintate that?

Something like

```
const QueryCondition = union(enum) {
    const Field = struct {}
    const ParaenthesisCondition = struct { /* uses QueryCondition here*/ }

    field: Field,
    parenthesis_condition: ParaenthesisCondition,
}
```

just so that you can immediately see "ah, those things are circular, so what I am looking at is a tree". 

I guess another nit is that `ParenthesisCondition` is perhaps not the best name, as its simultaneously long and is not very illuminating. 

But my bigger worry is that such recursive code feels un-TigerStyle. It says **Do not use recursion**, but `scan_from_condition` is recursive. While this recursive AST structure is near and dear to my heart of recovering compiler writer, I am wondering if there's some more TigerStyle representation here? 

For example, I think reverse polish notation might be a more conductive to simple loops?

```
const Query = BoundedArray(QueryPart, query_len_max);
const QueryPart = union(enum) {
    field: struct {index: Index, value: u64 }
    merge: struct { operator: QueryOperator, operand_count: u8 } // NB: non-recursive
}
```

then, `scan_from_condition` can go non-recursively as:

```
const stack = BoundedArray(*Scan, query_len_max);
for (query.items) |query_part| {
    switch(query_part) {
        .field => stack.push(scan_builder.scan_prefix())
        .merge => |merge| {
              const scan = scan_builder.merge(stack.items[stack.items.len - merge.operand_count..]);
              stack.pop_many(merge.operand_count);
              stack.push(scan);
        }
    }
}
assert(stack.len == 1)
```

Reply:
To clarify, I think its fine for fuzzers to be less TigerStyle than the main code. Still, perhaps it'd be interesting to use the fuzzer as a test bed to figure out patters for working with "trees" in TigerStyle. Maybe recursion is the best answer after all!